src/control/server/ctl_smd_rpc.go

```diff
@@ -92,23 +92,37 @@ func (svc *ControlService) querySmdDevices(ctx context.Context, req *ctlpb.SmdQu
 			}
 		}
 
-		if !req.IncludeBioHealth {
-			continue
-		}
-
+		i := 0 // output index
 		for _, dev := range rResp.Devices {
-			/* Skip health query if the device is in "NEW" state */
-			if storage.NvmeDevState(dev.DevState).IsNew() {
-				continue
+			state := storage.NvmeDevState(dev.DevState)
+
+			if req.ShowOnlyFaulty && !state.IsFaulty() {
+				continue // Skip device completely
 			}
```

Comment:
I'm not convinced that this filtering should be done on the server side. Generally the aim is to have relatively simple/thin RPC handlers and do most of the display-oriented filtering and logic on the client side. I can see that there is a potential performance benefit here, though, so I'm not strongly against it.

That having been said, if you're going to do the filtering on the server side, it seems like it would be better to make it more flexible and future-proof, perhaps by sending a bitmask to be used for matching devices in the desired states. The command surface can keep the bool flags for UI simplicity, though.

Reply:
We've had this conversation before, it's difficult to do filtering on the client side because of the way we handle host storage responses. I agree it's not ideal but it's the design we have.

I will send a state bitmask as you say, sounds sensible.